---
title: "genetic-map-construction"
author: "Ruijuan Li"
date: "5/7/2018"
output: html_document
---

### install & load package 
```{r}
library(tcltk)
library(tkrplot)
library(onemap)  
library(ggplot2)
library(reshape) 
library(tidyverse) 
library("qtl") 
library(snowfall) 
```

### import & format data

* input data are deposited in box folder https://ucdavis.app.box.com/folder/49886630432 input folder; output data are in  https://ucdavis.app.box.com/folder/49886630432 input folder 
```{r} 
F2_geno_data <- read.table("F2_Final_SNP_Calls", header = T)
rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)] 
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data)) 

F2_geno_data_with_missing_rate <- mutate(F2_geno_data, 
       missing_rate = round(apply(F2_geno_data, 1, function(x) sum(is.na(x)))/166, 2)) 

rownames(F2_geno_data_with_missing_rate) <- rownames(F2_geno_data) 
```

### remove markers with missing rate greater than 0.90
```{r}
F2_geno_data_with_missing_rate_filtered <-  F2_geno_data_with_missing_rate[F2_geno_data_with_missing_rate$missing_rate < 0.10,]

F2_geno_data_with_missing_rate_filtered.final <- subset(F2_geno_data_with_missing_rate_filtered, select=-missing_rate)
```

### calculate pairwise corrlelation and remove SNPs with corrlelation greater or equal to 0.9
some SNPs should have exactly the same genotype across all individuals, they are redundant in terms of genetic map construction, so they should be removed. find those SNPs by doing correlation test.
```{r}
F2_geno_data_t <- as.data.frame(t(F2_geno_data_with_missing_rate_filtered.final))
F2_geno_data_t.numeric <- data.matrix(F2_geno_data_t)

# delete markers with all same genotypes across individuals 
test <- apply(F2_geno_data_t.numeric, 2, function(x) length(unique(x[!is.na(x)])))
filter.polymorphsm <- test != 1 

F2_geno_data_t.numeric.2 <- F2_geno_data_t.numeric[,filter.polymorphsm]
dim(F2_geno_data_t.numeric.2) # 166 3606

# also output save markers with same genotype across individuals
polymorphism <- test == 1
SNP.not.poly <- colnames(F2_geno_data_t.numeric)[polymorphism]
length(SNP.not.poly) # two markers are not polymorphism across individuals 

# correlation test
options(warn=-1) # suppress warning message
F2_SNP_correlation <- cor(F2_geno_data_t.numeric.2, use = "pairwise.complete.obs")
options(warn=0) # unsuppress warning message
dim(F2_SNP_correlation) # 3606 3606 

# calcualte number of SNPs with correlation of 1
nrow(which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = T)) # 172 
# find SNPs with correlation of 1 and remove them 
dup.cordinate <- which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = F)
dup.cordinate.df <- as.data.frame(dup.cordinate)
sample.ID <- colnames(F2_SNP_correlation)

# extract duplicate pair information based on their coordicate
dup.pair <- data.frame(matrix(nrow = nrow(dup.cordinate), ncol = 2))
for (i in 1:nrow(dup.cordinate)){
 dup.pair[i,1] <- sample.ID[dup.cordinate[i,1]]
 dup.pair[i,2] <- sample.ID[dup.cordinate[i,2]]
}

rem <- unique(dup.pair[,2])
length(rem) # 163 markers are removed due to pairwise correlation of 1 

F2_geno_data_1 <- F2_geno_data_with_missing_rate_filtered.final[!(rownames(F2_geno_data_with_missing_rate_filtered.final) %in% rem),]  

# remove non polymorphism SNPs 
F2_geno_data_2 <- F2_geno_data_1[!(rownames(F2_geno_data_1) %in% SNP.not.poly),]

dim(F2_geno_data_2) # 3443  166, end up with 2932 SNPs for map construction 
```

### format data for onemap, polarize using parental genotype data 
```{r}
# read parent data 
Ae_Ol <- read.csv("vcf.Ae.Ol.intersect.final.csv", stringsAsFactors = F)

# left join to filter Ae_Ol SNP based on F2 genotypes
Ae_Ol$index <- paste(Ae_Ol$CHROM, Ae_Ol$POS, sep = "_")
F2_geno_data_2$index <- rownames(F2_geno_data_2)
F2_geno_data_2_Ae_Ol <- 
left_join(F2_geno_data_2, Ae_Ol, by="index") %>% 
  dplyr::select(-(X:ALT)) 

F2_geno_data_2_Ae_Ol <- as.matrix(F2_geno_data_2_Ae_Ol) 

# reassign genotype according to parents genotypes
F2_geno_data_2_Ae_Ol_new <- data.frame()

for (i in colnames(F2_geno_data_2_Ae_Ol)[1:166]) {
  print(i)
  for (j in 1:nrow(F2_geno_data_2_Ae_Ol)){
    if (is.na(F2_geno_data_2_Ae_Ol[j,i])){
    F2_geno_data_2_Ae_Ol_new[j,i] = "-"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == "0/1"){
      F2_geno_data_2_Ae_Ol_new[j,i] = "H"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == F2_geno_data_2_Ae_Ol[j,"Ae.gt"]){
      F2_geno_data_2_Ae_Ol_new[j,i] = "A"
    } else {
      F2_geno_data_2_Ae_Ol_new[j,i] = "B"
    }
  }
}

rownames(F2_geno_data_2_Ae_Ol_new) <- F2_geno_data_2_Ae_Ol[,"index"]
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166  

write.table(F2_geno_data_2_Ae_Ol_new, file = "F2_geno_data_2_Ae_Ol_new.txt")
# change file format in linux 
# cat F2_geno_data_2_Ae_Ol_new.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' | sed 's/ //g' > tmp
# tail -3443 tmp > tmp.1  

write.table(rownames(F2_geno_data_2_Ae_Ol_new), file = "marker_info.txt" )
# cat marker_info.txt | awk '{print "*"$2}' | tail -3334 | sed 's/"//g' > marker_info_reform.txt
# paste marker_info_reform.txt tmp.1 | awk '{print $1 " " $2}' > F2_geno_for_one_map.txt 
# cat header_one_map_input F2_geno_for_one_map.txt > F2_geno_for_one_map_final.txt 
# change header info: maker number to the right marker number 
```

### calculate two point rf 
https://github.com/MaloofLab/Li-eQTL-2018/blob/master/scripts/F2_map_construction_LOD3_rf0.5_missing_0.10.R   

### use madmapper to assign markers to clusters/LGs 
[http://cgpdb.ucdavis.edu/XLinkage/MadMapper/Genetic_Map_MadMapper_Arabidopsis.html]

### running madmapper 
### 1) format marker data to madmapper accepted format (on MAC)
```{r}
# cat F2_geno_data_2_Ae_Ol_new.txt | sed 's/"//g' | tr " " "\t" > F2_geno_data_m.txt 
# vim edit add ";" to the 1st line
# mv F2_geno_data_m.txt F2_geno_data_m.loc 
# wc -l F2_geno_data_m.loc 3444 
# cat F2_geno_data_m.loc | awk '{print NF}' | sort | uniq 
```

### 2) madmapper 
http://cgpdb.ucdavis.edu/XLinkage/MadMapper/#Part_2
```{r} 
# https://github.com/leejimmy93/KIAT/blob/master/madmapper.sh
# the output file we should check is tree.clust file 
```

check madmapper output (export *.tree.clust file to Excel, and decide which SNP belong to which LG based on tree value) 

```{r}
madmapper.result <- read.csv("F2_geno_data_m_loc_mmout.x_tree_clust.csv", header = F)

LG.A01 <- madmapper.result$V26[1:212] # 212 
LG.C01 <- madmapper.result$V26[225:355] # 131
LG.A03 <- madmapper.result$V26[356:621] # 266
LG.C03 <- madmapper.result$V26[659:844] # 186
LG.A09 <- madmapper.result$V26[845:1186] # 342
LG.C08 <- madmapper.result$V26[1187:1248] # 62
LG.A06 <- madmapper.result$V26[1261:1699] # 439
LG.A05 <- madmapper.result$V26[1717:1871] # 155
LG.C05 <- madmapper.result$V26[1872:1909] # 38
LG.A02 <- madmapper.result$V26[1924:2015] # 92
LG.A04 <- madmapper.result$V26[2016:2178] # 163
LG.C04 <- madmapper.result$V26[2179:2335] # 157
LG.A07 <- madmapper.result$V26[2336:2626] # 291
LG.A10 <- madmapper.result$V26[2627:2921] # 295
LG.C06 <- madmapper.result$V26[2922:3026] # 105
LG.C09 <- madmapper.result$V26[3027:3062] # 36
LG.A08 <- madmapper.result$V26[3075:3251] # 177
LG.C02 <- madmapper.result$V26[3252:3319] # 68
LG.C07 <- madmapper.result$V26[3320:3442] # 123

LG.madmapper <- list(LG.A01, LG.A02, LG.A03, LG.A04, LG.A05, LG.A06, LG.A07, LG.A08, LG.A09, LG.A10, LG.C01, LG.C02, LG.C03, LG.C04, LG.C05, LG.C06, LG.C07, LG.C08, LG.C09) 
length(LG.madmapper) # 19 

LG.C05 <- grep("chrC05", madmapper.result$V26, value = T) 
LG.C08 <- grep("chrC08", madmapper.result$V26, value = T) 

### get index for the above markers in different linkage group 
group.AC <- list()

for (i in 1:length(LG.madmapper)) {
  group.AC[[i]] <-  which(twopts.f2.LOD3_rf0.5$marnames %in% LG.madmapper[[i]])
}
save(group.AC, file = "group.AC.Rdata")
```  

#### order within each chromosome 
https://github.com/leejimmy93/KIAT/blob/master/F2/map_AC.R  

```{r}
# get order for A sub
LG.f2 <- list() 

for (i in c(1, 3, 4, 6, 7, 8, 9, 10, 11, 13)){
  LG.f2[[i]] <- make.seq(LG.f2.ord.A[[i]], "safe") 
}

for (i in c(1, 2, 3, 4, 6, 8, 9, 10, 12)){ 
  LG.f2[[13+i]] <- make.seq(LG.f2.ord.C[[i]], "safe") 
} 

# remove null LGs 
LG.f2 = LG.f2[-which(sapply(LG.f2, is.null))] 
draw.map(LG.f2, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2, "LG.f2.madmapper.C05C08.map")   
```

replace markers with double crossover (indication of mis-genotyping) with missing data, and remove markers with too many missgenotyping 
```{r}
## all markers on A & C 
LG.f2.madmapper <- read.cross("mm", file = "F2_geno_for_one_map_final.txt", mapfile = "LG.f2.madmapper.C05C08.map") 

summary(LG.f2.madmapper) # 2029  
plotMap(LG.f2.madmapper) 
totmar(LG.f2.madmapper)  

# iterate over all LGs 
# rename to the right name 
names(LG.f2.madmapper$geno)[1:10] <- paste("A", c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10"), sep = "")
names(LG.f2.madmapper$geno)[11:19] <- paste("C", c("01", "02", "03", "04", "05", "06", "07", "08", "09"), sep = "") 

# remove double crossover 
summaryMap(LG.f2.madmapper) # 6835
LG.f2.madmapper.before.crossover <- LG.f2.madmapper 

for (chr in names(LG.f2.madmapper$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2.madmapper$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2.madmapper$geno <- within(LG.f2.madmapper$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2.madmapper$geno)$data))))
}  

# restimate map with double crossover removed 
f2.map.new <- est.map(LG.f2.madmapper,verbose=T,error.prob=.01)  
LG.f2.madmapper <- replace.map(LG.f2.madmapper, f2.map.new)
LG.f2.madmapper.after.crossover <- LG.f2.madmapper
plot.map(LG.f2.madmapper.after.crossover, alternate.chrid = T) # the old genetic map

summaryMap(LG.f2.madmapper.after.crossover) # 2045

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/map_after_ripple_rmdoublecrossover_force.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper.after.crossover, col.scheme = "redblue", alternate.chrid = T) 
dev.off()

### deal with A10 
plot.rf(LG.f2.madmapper.after.crossover, chr = "A10")  
plotMap(LG.f2.madmapper.after.crossover, chr = 'A10') 

set.seed(16)
LG.f2.madmapper.after.crossover <- orderMarkers(LG.f2.madmapper.after.crossover, chr = "A10", 
	                        window = 5, use.ripple = TRUE, maxit = 4000, 
	                        error.prob = 0.0001, verbose = T)

plotMap(LG.f2.madmapper.after.crossover, chr = 'A10') 
plot.rf(LG.f2.madmapper.after.crossover, chr = 'A10') # plot appears better but length increased... still decided to use this new order 
plotMap(LG.f2.madmapper.after.crossover)
summaryMap(LG.f2.madmapper.after.crossover) # 3063
save(LG.f2.madmapper.after.crossover, file =  "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.C05C08.Rdata")
```
