---
title: "Gene-expression-analysis" 
author: "Ruijuan Li"
date: "5/7/2018"
output: 
  html_document: 
    keep_md: yes
---
### load libs
```{r}
library(edgeR)
library(ggplot2)
library("DESeq2") 
library(lme4) 
library(lmerTest) 
library(reshape2)
library(tidyverse)
```

### read in data 
```{r} 
setwd("~/Desktop/F2_paper/submission/Li-eQTL-2018/scripts/") 
parent.read.count <- read.table("../input/read.count.tsv", header = T, check.names = F)
rownames(parent.read.count) <- parent.read.count[,1]
parent.read.count <- parent.read.count[,-1]  

# format data 
even_indexes<-seq(2,length(colnames(parent.read.count)),2)
parent.read.count.one <- parent.read.count[,even_indexes]
colnames(parent.read.count.one) <- sub("_2.fq","",colnames(parent.read.count.one),fixed = TRUE) 

# sample description  
sample_des <- read.csv("../input/parent_summary.csv")
sorted_sample_des <- sample_des[order(sample_des$SampleID),]
new_sample_ID <- paste(sorted_sample_des$Cultivar, sorted_sample_des$Stage, sorted_sample_des$rep, sep = "_")
colnames(parent.read.count.one) <- new_sample_ID 
``` 

### sample description & filtering 
```{r} 
parent.read.count.one <- parent.read.count.one[,colSums(parent.read.count.one) > 1000000]  
parent.read.count.one.sample<-data.frame(file=colnames(parent.read.count.one),
                             batch=factor(gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",colnames(parent.read.count.one))),  
                             genotype=factor(gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",colnames(parent.read.count.one))),	
                             stage=factor(gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",colnames(parent.read.count.one))),	
                             group=factor(gsub("(Da-Ae|Da-Ol-1)(_)([[:print:]]+)(_)(1|2|3)","\\1\\3",colnames(parent.read.count.one)))
)

# filter based on read count 
parent.read.count.one.small <- parent.read.count.one[rowSums(parent.read.count.one > 10) >= 3,]
```

### vst transformation, used later for heritability calculation 
```{r}
dds.parent <- DESeqDataSetFromMatrix(countData = round(parent.read.count.one.small), colData = parent.read.count.one.sample, design = ~ batch + genotype*stage)

vsd.parent <- varianceStabilizingTransformation(dds.parent)
vstMat.parent <- assay(vsd.parent)
colnames(vstMat.parent) <- colnames(parent.read.count.one) 
```

### normalize & clustering  
```{r}
dge.new <- DGEList(counts=parent.read.count.one.small, group=parent.read.count.one.sample$group)
dge.new <- calcNormFactors(dge.new, method = "TMM") 
mds <- plotMDS(dge.new, method = "bcv",labels = dge.new$samples$group)

x <- as.data.frame(mds$x)
y <- as.data.frame(mds$y)
distance_matrix <- merge(x, y, by="row.names")
distance_matrix$group <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1\\3",distance_matrix$Row.names)
distance_matrix$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",distance_matrix$Row.names)
distance_matrix$tissue <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",distance_matrix$Row.names)

colnames(distance_matrix) <- c("lib","x","y","group","gt","tissue")

p.mds <- ggplot(data = distance_matrix) + theme_gray(base_size = 20) + 
  geom_point(aes(x, y, color=factor(gt), shape=factor(tissue)), size=5) + 
  labs(y = "BCV distance 2", x="BCV distance 1") + 
  theme()

p.mds
```

### design matrix and expression analysis 
```{r}
parent.read.count.one.sample$genotype <- as.factor(parent.read.count.one.sample$genotype)
parent.read.count.one.sample$tissue <- as.factor(parent.read.count.one.sample$stage)
parent.read.count.one.sample$genotype <- relevel(parent.read.count.one.sample$genotype,ref="Da-Ol-1")
parent.read.count.one.sample$tissue <- relevel(parent.read.count.one.sample$tissue,ref="Young")

design.new <- model.matrix(~tissue*genotype,data = parent.read.count.one.sample) 

# calculate dispersion
dge.new <- estimateGLMCommonDisp(dge.new, design.new,verbose = TRUE) 
dge.new <- estimateGLMTrendedDisp(dge.new,design.new)
dge.new <- estimateGLMTagwiseDisp(dge.new,design.new)

fit.new <- glmFit(dge.new, design.new)
lrt.new.interaction <- glmLRT(fit.new,coef = c("tissuebolting:genotypeDa-Ae", "tissueearly-silique:genotypeDa-Ae", "tissueflowering:genotypeDa-Ae", "tissuelate-silique:genotypeDa-Ae"))

DEgene.new.interaction <- topTags(lrt.new.interaction,n = Inf)$table[topTags(lrt.new.interaction,n = Inf)$table$FDR<0.05,]
nrow(DEgene.new.interaction)  

# genes for gt 
lrt.new.gt <- glmLRT(fit.new,coef = c("genotypeDa-Ae", "tissuebolting:genotypeDa-Ae", "tissueearly-silique:genotypeDa-Ae", "tissueflowering:genotypeDa-Ae", "tissuelate-silique:genotypeDa-Ae"))
DEgene.new.gt <- topTags(lrt.new.gt,n = Inf)$table[topTags(lrt.new.gt,n = Inf)$table$FDR<0.05,]
nrow(DEgene.new.gt)  
```

### GO enrichment analysis 
```{r}
source("helper.R") 
# gt effect  
DEgene.GO.ORA.gt <- GOseq.Bn.ORA(rownames(DEgene.new.gt))
DEgene.GO.ORA.interaction <- GOseq.Bn.ORA(rownames(DEgene.new.interaction))

# draw heatmap for gt & gt:tissue effect genes 
gt <- DEgene.GO.ORA.gt[,c("Term", "over_represented_padjust")] 
gt.tissue <- DEgene.GO.ORA.interaction[,c("Term", "over_represented_padjust")]
gt_gt.tissue <- merge(gt, gt.tissue, by="Term", all=TRUE)
names(gt_gt.tissue)[c(2:3)] <- c("genotype_effect", "genotype_by_tissue_effect")
gt_gt.tissue.melt <- melt(gt_gt.tissue)
gt_gt.tissue.melt$logPvalue <- -log10(gt_gt.tissue.melt$value)

# plot 
pl.heatmap1 <- ggplot(data = gt_gt.tissue.melt) + 
  geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=("royalblue")) + 
  labs(y = "GO term", x="", title=" ") + 
  theme_classic() + 
  theme(text = element_text(size=8), axis.text=element_text(size=rel(0.8))) 

pl.heatmap1    
```

### overlaps between genes with eQTL and genes differentially expressed between Da-Ae and Da-Ol-1 
```{r}
load("../output/cis_trans_result_new_flipped_C05C08.Rdata") 
eQTL_gene <- c(cis_eQTL$gene_ID, trans_eQTL$gene_ID) %>% unique() 
intersect(eQTL_gene, rownames(DEgene.new.gt)) %>% length() # 9436 
nrow(DEgene.new.gt) # 15098 
length(eQTL_gene) # 22693 

data <- matrix(c(9436, 5662, 13230, 42952), nrow = 2, ncol = 2)
colnames(data) <- c("DE", "no_DE")
rownames(data) <- c("eQTL", "no_eQTL")
data

fisher.test(data)  

# late silique only 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/output/parent/DEgene.pairwise.gt.Rdata")
nrow(DEgene.late.silique) # 7413 
intersect(eQTL_gene, rownames(DEgene.late.silique)) %>% length() # 5650 
unique(c(eQTL_gene, rownames(DEgene.late.silique))) %>% length() 

# genes with detectable expression in F2: 56180  
56180-  24456
data2 <- matrix(c(5650, 1763, 17043, 31724), nrow = 2, ncol = 2)
colnames(data2) <- c("DE", "no_DE")
rownames(data2) <- c("eQTL", "no_eQTL")
data2 

fisher.test(data2)  
```

### heritability calculation for all genes 

the reviwer challenged our experiment design, since our population is F2 population, which is impossible to have replicates, so the gene expression quantification could be inaccurate (envrionmental variation was not taken into consideration). To check for this problem, we decided to calculate the heritability of genes. The approach I will be using is very similar to what I have done for the phenotypic traits, but this will be done for all the genes. 

```{r, eval=FALSE}
# 1) get late silique gene quantification files for parents, vst transform the data, and calculate VE 
vstMat.parent %>% dim() # 60975    27 

# 2) F2 vst transformed data 
vstMat.f2.batch.corrected <- read.csv("../output/vstMat.f2.batch.corrected_revised.csv")
dim(vstMat.f2.batch.corrected) # 56180   167 

# 3) calculate VE using biological replicates of parent data  
vstMat.parent.late.silique <- 
vstMat.parent[rownames(vstMat.parent) %in% vstMat.f2.batch.corrected$id,] %>% 
  as.data.frame() %>% 
  dplyr::select(contains("late")) 

vstMat.parent.late.silique$PHENOTYPE <- rownames(vstMat.parent.late.silique) 

vstMat.parent.late.silique <- 
vstMat.parent.late.silique %>%
  gather(contains("Da"), key = "variable", value = "contains\\(\"Da\"\\)") %>% 
  mutate(genotype = gsub("(Da-Ol-1|Da-Ae)(_)(late-silique)(_)(1|2|3)", "\\1", variable), 
         rep_ID = gsub("(Da-Ol-1|Da-Ae)(_)(late-silique)(_)(1|2|3)", "\\5", variable))  

colnames(vstMat.parent.late.silique)[3] <- "value"

final_parent <- 
vstMat.parent.late.silique %>% 
  dplyr::select(genotype, rep_ID, PHENOTYPE, value) 

VarE <- 
sapply(unique(final_parent$PHENOTYPE), function(i) {
  print(final_parent[final_parent$PHENOTYPE == i,] %>% rownames() %>% tail(1))
  
  test <- final_parent %>% 
  filter(PHENOTYPE == i) 
  
  lme <- lmer(as.numeric(value) ~ (1|genotype),data=test)
  var <- as.data.frame(VarCorr(lme)) 
  var$vcov[2]
})

VarE <- data.frame(row.names = unique(final_parent$PHENOTYPE),
                   VarE = VarE) 
dim(VarE) # 51697     1 

geneID <- vstMat.f2.batch.corrected[vstMat.f2.batch.corrected$id %in% rownames(vstMat.parent),]$id

phefile.all <- 
vstMat.f2.batch.corrected[vstMat.f2.batch.corrected$id %in% rownames(vstMat.parent),] %>% 
  dplyr::select(-id) %>% 
  t() %>% 
  as.data.frame() 

colnames(phefile.all) <- geneID
phefile.all[1:10, 1:10] 

VarP <- 
sapply(colnames(phefile.all), function(i) var(phefile.all[[i]], na.rm = T)) %>% 
  as.data.frame()

colnames(VarP) <- "VarP" 
dim(VarP) # 51697     1

heritability <- 
merge(VarE, VarP, by = 0) %>% 
  mutate(heritability = (VarP - VarE)/VarP) %>% 
  arrange(desc(heritability)) 

heritability$heritability <- round(heritability$heritability, digits = 2)
# save(heritability, file = "../output/heritability.Rdata")
heritability$heritability <- ifelse(heritability$heritability < 0, 0, heritability$heritability) # heritability < 0 should be == 0 
```

```{r}
load("../output/heritability.Rdata")

# look for heritability for genes with eQTL and w/o eQTL  
eQTL_gene %>% length() # 22693 

heritability$eQTL <- ifelse(heritability$Row.names %in% eQTL_gene, "eQTL", "no_eQTL")

heritability_eQTL <- 
heritability %>% 
  ggplot() + 
  geom_histogram(aes(x = heritability), binwidth = 0.05) + 
  facet_wrap(~eQTL, nrow = 2) 

heritability_eQTL
# ggsave(heritability_eQTL, filename = "~/Desktop/F2_paper/submission/Li-eQTL-2018/output/heritability_eQTL.png", width = 6, height = 9) 
```


