Title: B subgenome check 
========================================================

Because Da-Ol-1 has introgression from B subgenome, we want to use our RNA-seq data to test this. 

The idea is that if there is introgression from B subgenome, this will present as a chunk of genes. So reads that map to A or C genome will map to B subgeome instead. And in the final B subgnome mapping result, there will be spikes where the read depth increase a lot.  

The way to test this is: map Da-Ol-1 reads to B.napus and B.napus + B subgenome (using STAR), get read depth for a sliding window of 1000bp (samtool or bedtools), then to see whether there are read preferentially map to B subgenome, for each sliding window mapping position on A + C genome, plot the differences in read depth between the two mapping result. If there are differences, plot the read depth on B subgenome only, and we expect to see a cluster of sliding window positions with high read depth.  

### get B.napus + B subgenome 
```{r}
# 1) get B subgenome from B.juncea 
# J11-J18 in B.juncea are B subgenome 
# perl -ne 'if(/^>(\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' Bsub_ID Bju_genome.fa > Bju_B_genome
# cat Brassica_napus_v4.1.chromosomes.fa Bju_B_genome > B_napus_plus_Bsub.fa

# 2) generate STAR genome index w/o annotation file (gff)
# mkdir star_genome 
# STAR --runMode genomeGenerate --genomeDir star_genome/ --genomeFastaFiles B_napus_plus_Bsub.fa --runThreadN 6 

# 3) map 3 Da-Ol-1 libraries to this synthesized hexaploid genome 
# https://github.com/leejimmy93/KIAT/blob/master/mapping_Ol.sh 
# by checking the Log.final.out file, the unique mapping ratio dropped a little bit, a little bit increase of total mapping ratio, no big difference though... (maybe need figure to demonstrate this...)
```

### For paper revision 

After paper submission to Frontiers in Plant biologiy, we were challenged by a Q from a reviewer: Could there be fragments from B-genome? Did you try to map your reads against the B genome? 

Since I just deleted this folder, so I will re-run everything from the start. I can re-run just the flower data, 3 samples: 2, All1_Gae_2, All1_Gae_3  

1) map flower data from Da-Ol-1 to B.napus & B.napus + B subgenome, 0 mismatch for both mapping_napus.sh & mapping_napus_Bsub.sh

checking mapping result 
```{r}
mapping.ABC <- read.table("~/B_subgenome_check/output/mapping_result/ABC/Star_Stats.tab", header=T)
mapping.AC <- read.table("~/B_subgenome_check/output/mapping_result/AC/Star_Stats.tab", header=T)

head(mapping.ABC)
head(mapping.AC)

mapping.AC$Sample <- gsub("_paired.star.trim.dir", "", mapping.AC$Sample)
mapping.ABC$type <- rep("ABC", 1)
mapping.AC$type <- rep("AC", 1)

mapping <- rbind(mapping.ABC, mapping.AC)
mapping.sub <- mapping[,c("Sample", "Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short", "type")]
mapping.sub.melt <- melt(mapping.sub)
mapping.sub.melt

p.mapping <- ggplot(data=mapping.sub.melt)
p.mapping <- p.mapping + geom_col(aes(x=variable, y=value, fill=variable))
p.mapping <- p.mapping + facet_grid(type~Sample)
p.mapping <- p.mapping + geom_text(aes(x=variable, y=value, label=value))
p.mapping <- p.mapping + theme(axis.text.x = element_text(angle = 90, size = 8))
p.mapping
ggsave(p.mapping, filename="~/B_subgenome_check/output/figure/mapping.png", height=8, width=11)
```

2) extract uniquely mapped reads from both mapping result 
https://github.com/leejimmy93/KIAT/blob/master/parent/extract_unique_mapping_reads.sh
https://github.com/leejimmy93/KIAT/blob/master/parent/extract_unique_mapping_reads_Bsub.sh  

3) calculate read depth for each position  
https://github.com/leejimmy93/KIAT/blob/master/parent/check_B_subgenome_Da_Ol.sh
https://github.com/leejimmy93/KIAT/blob/master/parent/check_B_subgenome_Da_Ol_2.sh 

4) get the difference in read depth between ABC VS AC mapped reads in 10,000 bp window size, next time should do average before load into R, can do average using python script  

```{r}
ABC_A01 <- read.table("~/B_subgenome_check/output/mapping_result/ABC/2_Unique.sorted.bam_chrA01.depth")
AC_A01 <- read.table("~/B_subgenome_check/output/mapping_result/AC/2_Unique.sorted.bam_chrA01.depth")
data <- prep.4.plot.depth.diff(depth_AC=AC_A01, depth_ABC=ABC_A01) 
plot(data, xlab="A01", ylim=c(-4e+5, 8e+5), ylab="") 

J11 <- read.table("~/B_subgenome_check/output/mapping_result/ABC/2_Unique.sorted.bam_J11.depth")
J11_in_window <- total.by.window.size(J11) 
plot(J11_in_window, xlab="J11", ylim=c(0, 8e+5), ylab="")

path_ABC <- "~/B_subgenome_check/output/mapping_result/ABC/depth"
files <- list.files(path=path_ABC)
files <- grep("depth", files, value = T)
setwd(path_ABC)
ABC_depth <- lapply(files, function(x) read.table(x)) 
files <- gsub("Unique.sorted.bam_", "", files)
files <- gsub("\\.depth", "", files)
names(ABC_depth) <- files

path_AC <- "~/B_subgenome_check/output/mapping_result/AC/depth"
files <- list.files(path=path_AC)
files <- grep("depth", files, value = T)
setwd(path_AC)
AC_depth <- lapply(files, function(x) read.table(x))   
files <- gsub("Unique.sorted.bam_", "", files)
files <- gsub("\\.depth", "", files)
names(AC_depth) <- files

# average across 10k bp and plot 
# all J 
ABC_depth_B <- ABC_depth[grep("J", names(ABC_depth), value = T)] 
tmp <- total.by.window.size(ABC_depth_B[1:8]) 
ABC_depth_B_window_average <- lapply(names(ABC_depth_B[1:8]), function(i) total.by.window.size((ABC_depth_B[[i]])))
names(ABC_depth_B_window_average) <- names(ABC_depth_B[1:8])

png("~/B_subgenome_check/output/figure/Ol_2_J_depth.png", width=10, height=20, units="in", res=300)
par(mfrow=c(8,1)) 
for(i in names(ABC_depth_B_window_average)){
  plot(ABC_depth_B_window_average[[i]], xlab=i, ylim=c(0, 8e+5), ylab="") 
}
dev.off()

ABC_depth_diff_window_average <- 
lapply(names(AC_depth[1:19]), function(i) {
  prep.4.plot.depth.diff(depth_AC=AC_depth[[i]], depth_ABC=ABC_depth[[i]]) 
}) 
names(ABC_depth_diff_window_average) <- names(AC_depth[1:19])

png("~/B_subgenome_check/output/figure/ABC_VS_AC_depth.png", width=12, height=20, units="in", res=300)
par(mfrow=c(10,2)) 
for(i in names(ABC_depth_diff_window_average)){
  plot(ABC_depth_diff_window_average[[i]], xlab=i, ylim=c(-4e+5, 8e+5), ylab="") 
}

dev.off() 
```

### run the same pipeline for Da-Ae, as a negative control 
mapping result 
```{r}
mapping.ABC <- read.table("~/B_subgenome_check/output/mapping_result/ABC/Star_Stats.tab", header=T)
mapping.AC <- read.table("~/B_subgenome_check/output/mapping_result/AC/Star_Stats.tab", header=T)

head(mapping.ABC)
head(mapping.AC)

mapping.AC$Sample <- gsub("_paired.star.trim.dir", "", mapping.AC$Sample)
mapping.ABC$type <- rep("ABC", 3)
mapping.AC$type <- rep("AC", 3)

mapping <- rbind(mapping.ABC, mapping.AC)
mapping.sub <- mapping[,c("Sample", "Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short", "type")]
mapping.sub.melt <- melt(mapping.sub)
mapping.sub.melt

p.mapping <- ggplot(data=mapping.sub.melt)
p.mapping <- p.mapping + geom_col(aes(x=variable, y=value, fill=variable))
p.mapping <- p.mapping + facet_grid(type~Sample)
p.mapping <- p.mapping + geom_text(aes(x=variable, y=value, label=value))
p.mapping <- p.mapping + theme(axis.text.x = element_text(angle = 90, size = 8))
p.mapping
ggsave(p.mapping, filename="~/B_subgenome_check/output/figure/mapping.Ae.png", height=8, width=11) 
```

### depth result 
```{r}
path_ABC <- "~/B_subgenome_check/output/mapping_result/ABC/depth/Ae/"
files <- list.files(path=path_ABC)
files <- grep("depth", files, value = T)
setwd(path_ABC) 
ABC_depth <- lapply(files, function(x) read.table(x)) 
files <- gsub("Unique.sorted.bam_", "", files)
files <- gsub("\\.depth", "", files)
names(ABC_depth) <- files 

path_AC <- "~/B_subgenome_check/output/mapping_result/AC/depth/Ae/"
files <- list.files(path=path_AC)
files <- grep("depth", files, value = T)
setwd(path_AC)
AC_depth <- lapply(files, function(x) read.table(x))   
files <- gsub("Unique.sorted.bam_", "", files)
files <- gsub("\\.depth", "", files)
names(AC_depth) <- files

# average across 10k bp and plot 
# all J 
ABC_depth_B <- ABC_depth[grep("J", names(ABC_depth), value = T)] 
ABC_depth_B_window_average <- lapply(names(ABC_depth_B), function(i) total.by.window.size((ABC_depth_B[[i]])))
names(ABC_depth_B_window_average) <- names(ABC_depth_B)

png("~/B_subgenome_check/output/figure/Ae_2_J_depth.png", width=10, height=20, units="in", res=300)
par(mfrow=c(8,1)) 
for(i in names(ABC_depth_B_window_average)){
  plot(ABC_depth_B_window_average[[i]], xlab=i, ylim=c(0, 8e+5), ylab="") 
}
dev.off()

ABC_depth_diff_window_average <- 
lapply(names(AC_depth), function(i) {
  prep.4.plot.depth.diff(depth_AC=AC_depth[[i]], depth_ABC=ABC_depth[[i]]) 
}) 
names(ABC_depth_diff_window_average) <- names(AC_depth)

png("~/B_subgenome_check/output/figure/Ae_ABC_VS_AC_depth.png", width=12, height=20, units="in", res=300)
par(mfrow=c(10,2)) 
for(i in names(ABC_depth_diff_window_average)){
  plot(ABC_depth_diff_window_average[[i]], xlab=i, ylim=c(-4e+5, 8e+5), ylab="") 
}

dev.off() 
```

### unique mapping rate to B-subgenome VS A&C 
```{r}
ABC_mapping <- read.table("~/Desktop/F2_paper/submission/Li-eQTL-2018/input/Stats.tab", header = T)

ABC_mapping <- 
ABC_mapping %>% 
  mutate(Sample = gsub("Unique.sorted.bam_", "", Sample)) %>% 
  mutate(Sample = gsub(".bam.stats", "", Sample)) %>% 
  mutate(lib_ID = gsub("(All1_Gae_3|All1_Gae_2|2|6)([[:print:]]+)", "\\1", Sample)) %>% 
  mutate(chr = gsub("(All1_Gae_3|All1_Gae_2|2|6)(_)([[:print:]]+)", "\\3", Sample)) %>% 
  dplyr::select(lib_ID, chr, Number_Unique_Mapped) 

p.ABC_mapping <- 
ABC_mapping[grep("random", ABC_mapping$chr, invert = T),] %>% 
  group_by(lib_ID) %>% 
  mutate(Total = sum(Number_Unique_Mapped)) %>% 
  mutate(Pct_uniq = Number_Unique_Mapped/Total) %>%
  mutate(chr = gsub("chr", "", chr)) %>% 
  ggplot() + 
  geom_col(aes(x = chr, y = Pct_uniq, fill = lib_ID)) + 
  facet_wrap(~lib_ID, nrow = 2, ncol = 2) + 
  theme(legend.position = "none")

ggsave(p.ABC_mapping, filename = "~/Desktop/F2_paper/submission/Li-eQTL-2018/output/p.ABC_mapping.png", width = 15, height = 3)  

ABC_mapping %>% 
  group_by(lib_ID) %>% 
  mutate(Total = sum(Number_Unique_Mapped)) %>% 
  mutate(Pct_uniq = Number_Unique_Mapped/Total) %>%
  mutate(sub = gsub("(chr)([[:print:]]+)", "\\1", chr)) %>% 
  mutate(sub = ifelse(sub == "chr", "AC", "B")) %>% 
  filter(sub == "B") %>% 
  group_by(lib_ID) %>% 
  summarise(B_sub_mapping = sum(Pct_uniq))
  
```



