---
title: "QTL_eQTL_mapping_2"
author: "Ruijuan Li"
date: "5/13/2018"
output: 
  html_document: 
    keep_md: yes
---

### load libraries 
```{r}
library(tidyverse)  
library(qtl) 
library(Biostrings)
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures) 
library("rtracklayer")
``` 

### load data 
```{r,eval=FALSE} 
library(boxr)
box_auth()
box_setwd(49886630432)
box_load(295501997932) # QTL result 
box_load(295501558109) # eQTL result
```

### QTL mapping result output 
```{r, eval=FALSE}
setwd("~/Desktop/F2_paper/submission/Li-eQTL-2018/scripts/")

### summarize QTL mapping result  
threshold.95 <- tibble(perm.threshold = bind_rows(scanone.perm.imp.all) %>% as.numeric(), 
                       trait = colnames(bind_rows(scanone.perm.imp.all)))

scanone.qtl.2 <-  
bind_cols(scanone.imp.all) %>% 
  dplyr::select(chr, pos, starts_with("lod"))
rownames(scanone.qtl.2) <- rownames(scanone.imp.all$Crude_oil_contents)
colnames(scanone.qtl.2)[3:ncol(scanone.qtl.2)] <- names(scanone.imp.all)

scanone.gather <- scanone.qtl.2 %>%
  gather(key = trait, value = LOD, -chr, -pos) %>%
  left_join(threshold.95)

sig.chrs <- scanone.gather %>% dplyr::filter(LOD > perm.threshold) %>%
  group_by(trait,chr) %>%
  dplyr::summarise(count = n())

# now for each significant chromosome/trait combo run bayesint
bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(scanone.qtl.2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE 
  )
  colnames(result)[3] <- "LOD" 
  result
})  
 
names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x)  
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr))
) 

bayesint.list %>% length() # 33
bayesint.list.scanone <- bayesint.list 

bayesint.result.scanone <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD), genetic_start=min(pos, na.rm = T), genetic_end=max(pos, na.rm = T)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(start=ifelse(start==end,max(0,start-20000),start), end=ifelse(start==end,end+20000,end))

bayesint.result.scanone %>% dim() # 33 8 

### cim
threshold.95 <- tibble(perm.threshold = bind_rows(cim.perm.all) %>% as.numeric(), 
                       trait = colnames(bind_rows(cim.perm.all)))

cim.qtl.2 <-  
bind_cols(cim.qtl.all) %>% 
  dplyr::select(chr, pos, starts_with("lod"))

rownames(cim.qtl.2) <- rownames(cim.qtl.all$Crude_oil_contents)
colnames(cim.qtl.2)[3:ncol(cim.qtl.2)] <- names(cim.qtl.all)

cim.gather <- cim.qtl.2 %>%
  gather(key = trait, value = LOD, -chr, -pos) %>%
  left_join(threshold.95)

# look for overlap, for each trait, find QTL border and look for genes under QTL peaks 
sig.chrs <- cim.gather %>% dplyr::filter(LOD > perm.threshold) %>%
  group_by(trait,chr) %>%
  dplyr::summarise(count = n()) 

# now for each significant chromosome/trait combo run bayesint
bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(cim.qtl.2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE 
  )
  colnames(result)[3] <- "LOD" 
  result
})  

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x)  
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr))
) 

bayesint.list %>% length() # 26 

# save bayesint result for later 
bayesint.list.cim <- bayesint.list

bayesint.result.cim <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD), genetic_start=min(pos, na.rm = T), genetic_end=max(pos, na.rm = T)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(start=ifelse(start==end,max(0,start-20000),start), end=ifelse(start==end,end+20000,end))

bayesint.result.cim %>% dim() # 26 8

bayesint.result.cim$model <- rep("cim", nrow(bayesint.result.cim))
bayesint.result.scanone$model <- rep("scanone", nrow(bayesint.result.scanone))

bayesint.result <- 
bayesint.result.cim %>% 
  full_join(bayesint.result.scanone, by = c("trait", "chr")) 

bayesint.result %>% dim() # 35 16 

# mark QTL that were detected from both methods 
note <- ifelse((rowSums(is.na(bayesint.result)) == 0), "*", "")

for (i in 1:nrow(bayesint.result)){
  if(is.na(bayesint.result[i, "start.x"])){
    bayesint.result[i, c(3:9)] <- bayesint.result[i, c(10:16)]
  }
}

bayesint.result <- bayesint.result[,1:9] 
colnames(bayesint.result) <- gsub("\\.x$", "", colnames(bayesint.result)) 

bayesint.result %>% dim() # 35 9 

# make Table 2
length(bayesint.list.scanone) # 33
length(bayesint.list.cim) # 26 

bayesint.result.scanone <- as.tibble(bind_rows(bayesint.list.scanone,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(pos, na.rm = T),end=max(pos, na.rm = T),pos = median(pos, na.rm = T), LOD=max(LOD))  

bayesint.result.scanone %>% dim() # 33 6 

bayesint.result.tmp <- as.tibble(bind_rows(bayesint.list.scanone,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  # separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) 

bayesint.result.tmp$index <- paste(bayesint.result.tmp$trait, bayesint.result.tmp$chr, bayesint.result.tmp$LOD)
bayesint.result.scanone$index <- paste(bayesint.result.scanone$trait, bayesint.result.scanone$chr, bayesint.result.scanone$LOD)

test <- 
bayesint.result.tmp %>% 
  anti_join(bayesint.result.scanone) %>% 
  dplyr::select(trait, chr, markername) %>% 
  mutate(index = paste(trait, chr, sep = "_")) 

tmp <- c()

for (i in seq_along(1:(nrow(test)/2))){
  tmp[i] <- paste(test$markername[i*2-1], test$markername[i*2-0], sep = "-") 
}

bayesint.result.scanone$index <- tmp

bayesint.result.scanone <- 
bayesint.result.scanone %>%
  mutate(start = round(start, 2), end = round(end, 2), pos = round(pos, 2), LOD = round(LOD, 2)) %>%
  unite(confidence_interval, start, end, sep = "-") %>% 
  mutate(flanking_marker = index) %>% 
  dplyr::select(-index)  

bayesint.result.scanone %>% dim() # 33 6 

### cim result 
bayesint.list.cim %>% length() # 26

bayesint.result <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(pos, na.rm = T),end=max(pos, na.rm = T),pos = median(pos, na.rm = T), LOD=max(LOD)) 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.

bayesint.result %>% dim() # 26 6 

bayesint.result.tmp <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  # separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) 

bayesint.result.tmp$index <- paste(bayesint.result.tmp$trait, bayesint.result.tmp$chr, bayesint.result.tmp$LOD)
bayesint.result$index <- paste(bayesint.result$trait, bayesint.result$chr, bayesint.result$LOD)

test <- 
bayesint.result.tmp %>% 
  anti_join(bayesint.result) %>% 
  dplyr::select(trait, chr, markername) %>% 
  mutate(index = paste(trait, chr, sep = "_")) 

tmp <- c()

for (i in seq_along(1:(nrow(test)/2))){
  tmp[i] <- paste(test$markername[i*2-1], test$markername[i*2-0], sep = "-") 
}

bayesint.result$index <- tmp 

bayesint.result.cim <- 
bayesint.result %>% 
  mutate(start = round(start, 2), end = round(end, 2), pos = round(pos, 2), LOD = round(LOD, 2)) %>%
  unite(confidence_interval, start, end, sep = "-") %>% 
  mutate(flanking_marker = index) %>% 
  dplyr::select(-index)  

bayesint.result.cim %>% dim() # 26 6  

### combine cim & scanone result 
bayesint.result.scanone$model <- rep("scanone", nrow(bayesint.result.scanone))
bayesint.result.cim$model <- rep("cim", nrow(bayesint.result.cim))

bayesint.result.paper <- 
bayesint.result.cim %>% 
  full_join(bayesint.result.scanone, by = c("trait", "chr")) 

bayesint.result.paper %>% dim() ### 35 12

for (i in 1:nrow(bayesint.result.paper)){
  if(is.na(bayesint.result.paper[i, "pos.x"])){
    bayesint.result.paper[i, c(3:7)] <- bayesint.result.paper[i, c(8:12)]
  }
}

bayesint.result.paper <- bayesint.result.paper[,1:7] 
colnames(bayesint.result.paper) <- gsub("\\.x$", "", colnames(bayesint.result.paper)) 

bayesint.result.paper %>% dim() # 35 7  
bayesint.result.paper$model <- paste(bayesint.result.paper$model, note, sep = "")

write.csv(bayesint.result.paper, file = "../output/bayesint.result.paper.csv") 

### combine with allele effect information, allelic effect was calculated using fitqtl(), and values were added to bayesint.result.paper.csv  
bayesint.result.allele_effect <- read.csv("../input/bayesint.result.paper_allele_effect.csv")

bayesint.result.paper.final <- 
bayesint.result.paper %>%
  left_join(bayesint.result.allele_effect, by = c("trait", "chr")) %>% 
  mutate(model = model.x) %>%
  dplyr::select(-model.x, -model.y, -X) 

### add physical position interval  
bayesint.result.scanone <- as.tibble(bind_rows(bayesint.list.scanone,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD))

bayesint.result.cim <- as.tibble(bind_rows(bayesint.list.cim,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD))

bayesint.result.cim$model <- rep("cim", nrow(bayesint.result.cim))
bayesint.result.scanone$model <- rep("scanone", nrow(bayesint.result.scanone))

bayesint.result <- 
bayesint.result.cim %>% 
  full_join(bayesint.result.scanone, by = c("trait", "chr")) 

for (i in 1:nrow(bayesint.result)){
  if(is.na(bayesint.result[i, "start.x"])){
    bayesint.result[i, c(3:7)] <- bayesint.result[i, c(8:12)]
  }
}

bayesint.result <- bayesint.result[,1:7] 
colnames(bayesint.result) <- gsub("\\.x$", "", colnames(bayesint.result)) 

bayesint.result %>% dim() # 35 7 

bayesint.result.physical <-
bayesint.result.paper.final %>% 
  left_join(bayesint.result, by= c("trait", "chr")) %>% 
  mutate(model = model.x) %>% 
  dplyr::select(-min_eQTL_LOD, -max_eQTL_LOD, -model.y, -model.x) 
  
bayesint.result.physical 
write.csv(bayesint.result.physical, file = "../output/bayesint.result.paper.physical.csv") 

# annotate QTL   
load("../input/BnapusAnnotation.Rdata") 

traitQTL.annotated <- lapply(1:nrow(bayesint.result),function(row) { # for each trait/module 
  qtl <- bayesint.result[row,]  
  results <- subset(BnapusAnnotation, chrom==qtl$chr &
                    start >= qtl$start & # genes which fall into the QTL interval 
                    end <= qtl$end)
} 
)   

names(traitQTL.annotated) <- bayesint.result$trait 

traitQTL.annotated <- bind_rows(traitQTL.annotated,.id="trait") %>% # combine list into data.frame 
  mutate(chrom=as.character(chrom)) %>%
  left_join(bayesint.result,by=c("trait","chrom"="chr")) #get eQTL LOD

traitQTL.annotated <- 
traitQTL.annotated %>% 
  mutate(start = start.y, end = end.y) %>% 
  dplyr::select(-start.x, -end.x, -start.y, -end.y, -min_eQTL_LOD, -max_eQTL_LOD) 

traitQTL.annotated %>% dim() # 18647     11

## get GO term for each gene
load("../input/napus_GO_combined.Rdata") 

colnames(traitQTL.annotated)[3] <- "gene_ID"

traitQTL.annotated <- 
traitQTL.annotated %>% 
  left_join(napus_GO_combined) 

traitQTL.annotated %>% dim() #  18647    13 
save(traitQTL.annotated, file =  "../output/traitQTL.annotated.flipped.Rdata")  
```

### eQTL mapping result output  
```{r, eval=FALSE}
# determine cis- and trans- eQTL 
scanone_eQTL.F2 %>% dim() # 4944 56182 

scanone_eQTL.F2$chr <- as.character(scanone_eQTL.F2$chr) 

# get threshold 
threshold.95 <- lod.thrs[5,] 
threshold.95 # 4.18 

# get all eQTL based on this threshold   
eQTL_sign <- 
sapply(colnames(scanone_eQTL.F2), function(gene) {
  sum(scanone_eQTL.F2[,gene] > threshold.95) > 0 
})  

sum(eQTL_sign) # 22,695 genes with eQTL 
scanone_eQTL.F2 <- scanone_eQTL.F2[,eQTL_sign]  
dim(scanone_eQTL.F2) # 4944 22695  

# get bayesint result for every gene 
scanone.gather <-  
scanone_eQTL.F2 %>% 
  gather(key = trait, value = LOD, -chr, -pos) 

sig.chrs <- scanone.gather %>% filter(LOD > threshold.95) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarise(count = n()) # this is to get the significant chr ID for each trait 

sig.chrs %>% dim() # 26244         3

bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(scanone_eQTL.F2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE, 
                     prob = 0.999 # expand the marker interval for cis-eQTL determination
  )
  colnames(result)[3] <- "LOD" 
  result
})  

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x) 
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr)) 
)  

bayesint.result <- 
as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
    dplyr::select(trait,chr,pos,markername,LOD) %>%  
    separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>%  
    group_by(trait,chr) %>% 
    dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T), pos=median(pos, na.rm = T), min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD), genetic_start=min(pos, na.rm = T), genetic_end=max(pos, na.rm = T)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  # mutate(start=ifelse(start==end, start-20000,start), end=ifelse(start==end,end+20000,end)) 
  mutate(start_modified=ifelse(start-250000>0, start-250000, 0), end_modified=end+250000) 

bayesint.result %>% dim() # 26244    11

# get genome range 
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures) 
library("rtracklayer") 

### get gff file with gene chrom & pos info, gff3 file must be sorted 
gff.mRNA <- read.table("../input/gff.mRNA")
dim(gff.mRNA) # 101040      4 
colnames(gff.mRNA) <- c("gene_CHROM", "gene_start", "gene_end", "gene_ID") 

# look for cis-eQTL  
bayesint.result.2 <- 
bayesint.result %>% 
  mutate(gene_ID = trait, eQTL_chr = chr, eQTL_start = start_modified, eQTL_end = end_modified) %>% 
  dplyr::select(trait, gene_ID, eQTL_chr, pos, eQTL_start, eQTL_end, start, end, min_eQTL_LOD, max_eQTL_LOD, genetic_start, genetic_end) %>% 
  left_join(gff.mRNA, by = "gene_ID") 

bayesint.result.2$eQTL_chr <- paste("chr", bayesint.result.2$eQTL_chr, sep = "")

cis_eQTL <- 
bayesint.result.2 %>% 
  filter(eQTL_chr == gene_CHROM) %>% 
  filter((gene_start < eQTL_start & gene_end > eQTL_start) | # also need SNP pos... 
         (gene_start >= eQTL_start & gene_end <= eQTL_end) |
         (gene_start < eQTL_end & gene_end > eQTL_end)) 

dim(cis_eQTL) # 11031    15

trans_eQTL <- 
bayesint.result.2 %>% 
  anti_join(cis_eQTL) 

save(cis_eQTL, trans_eQTL, file = "../output/cis_trans_result_new_flipped_C05C08.Rdata") 
```

### major QTL plot 
```{r}
load("../input/QTL_result_all.Rdata")
load("../input/bayesint.result.physical.Rdata")

bayesint.result_scanone <- 
bayesint.result %>% 
  filter(model == "scanone")  

scanone_result <- list()

for (i in 1:nrow(bayesint.result_scanone)) {
  trait <- bayesint.result_scanone$trait[i] 
  chr1 <- bayesint.result_scanone$chr[i] 
  tmp <- scanone.imp.all[[trait]] 
  scanone_result[[i]] <- tmp %>% filter(chr == chr1) 
}

names(scanone_result) <- bayesint.result_scanone$trait

bayesint.result_cim <- 
bayesint.result %>% 
  filter(model == "cim")  

cim_result <- list()

for (i in 1:nrow(bayesint.result_cim)) {
  trait <- bayesint.result_cim$trait[i] 
  chr1 <- bayesint.result_cim$chr[i] 
  tmp <- cim.qtl.all[[trait]] 
  cim_result[[i]] <- tmp %>% filter(chr == chr1) 
}

names(cim_result) <- bayesint.result_cim$trait

# combine result and plot 
# add model label 
for(i in 1:length(scanone_result)){
  scanone_result[[i]]$model <- rep("scanone", nrow(scanone_result[[i]]))
  scanone_result[[i]]$trait <- rep(names(scanone_result)[i], nrow(scanone_result[[i]]))
}

for(i in 1:length(cim_result)){
  cim_result[[i]]$model <- rep("cim", nrow(cim_result[[i]]))
  cim_result[[i]]$trait <- rep(names(cim_result)[i], nrow(cim_result[[i]]))
}

scanone <- 
lapply(1:length(scanone_result), function(i) {
  tmp <- scanone_result[[i]] %>% 
    filter(chr == "A08" | chr == "C03")
  if(nrow(tmp) > 0){
    tmp
  }
})

cim <- 
lapply(1:length(cim_result), function(i) {
  tmp <- cim_result[[i]] %>% 
    filter(chr == "A08" | chr == "C03")
    if(nrow(tmp) > 0){
    tmp
    }
})

final_FA <- do.call(rbind, plyr::compact(c(scanone, cim))) 

final_FA$lod <- ifelse(log10(final_FA$lod) <= -2, 0.01, final_FA$lod)
p.final_FA <-
final_FA %>% 
  ggplot() +
  geom_tile(color = "white", aes(x = factor(pos), y = trait, fill=lod)) +  scale_fill_gradient2(low=("white"), mid = "white", high=("magenta"), midpoint = 1, trans = "log", breaks = c(0.5, 3, 20)) +   
  facet_grid(trait~chr, scales = "free") + 
  labs(y = "", x="genetic position", title="Fatty acid major QTL") + 
  theme_bw() +
  theme(text = element_text(size=8)) + 
  theme(axis.text.x=element_blank(), strip.text.y = element_blank(), plot.title = element_text(hjust = 0.5))   

# flower etc. 
scanone_2 <- 
lapply(1:length(scanone_result), function(i) {
  tmp <- scanone_result[[i]] %>% 
    filter(chr == "A10" | chr == "C06")
  if(nrow(tmp) > 0){
    tmp
  }
})

cim_2 <- 
lapply(1:length(cim_result), function(i) { 
  tmp <- cim_result[[i]] %>% 
    filter(chr == "A10" | chr == "C06")
    if(nrow(tmp) > 0){
    tmp
    }
})

final_flower_etc <- do.call(rbind, plyr::compact(c(scanone_2, cim_2))) 

p.flower_etc <-
final_flower_etc %>%  
  filter(trait != "Myristic_acid") %>% 
  ggplot() +
  geom_tile(color = "white", aes(x = factor(pos), y = trait, fill=lod)) + scale_fill_gradient2(low=("green"), high=("magenta"), midpoint = 1) +
  facet_grid(trait~chr, scales = "free") + 
  labs(y = "", x="genetic position", title="flowering time and growth-related traits major QTL") + 
  theme_bw() + 
  theme(text = element_text(size=8)) + 
  theme(axis.text.x=element_blank(), strip.text.y = element_blank(), plot.title = element_text(hjust = 0.5)) 

library(cowplot)

plot.all.paper <-plot_grid(
  p.final_FA,
  p.flower_etc,
  ncol=1, nrow = 2,labels=c("a","b"), label_size = 10)  

plot.all.paper 
```

### scantwo & epistatic effect for linolenic acid 

scantwo analysis on HPC https://github.com/MaloofLab/Li-eQTL-2018/tree/master/scripts/scantwo 
```{r} 
# linolenic acid scantwo result use effectplot() in Q/qtl package 
```

### plot eQTL result 
```{r}
setwd("~/Desktop/F2_paper/submission/Li-eQTL-2018/scripts/") 
load("../output/cis_trans_result_new_flipped_C05C08.Rdata") 

dim(cis_eQTL)
dim(trans_eQTL) 

# output plot 
cis_eQTL$index <- paste(cis_eQTL$eQTL_chr, cis_eQTL$eQTL_start, cis_eQTL$eQTL_end, sep = "_")
trans_eQTL$index <- paste(trans_eQTL$eQTL_chr, trans_eQTL$eQTL_start, trans_eQTL$eQTL_end, sep = "_")

cis_eQTL$class <- rep("cis", nrow(cis_eQTL))  
trans_eQTL$class <- rep("trans", nrow(trans_eQTL))   

# get only main chromosome stuff   
eQTL <- rbind(cis_eQTL, trans_eQTL) 
eQTL <- eQTL[grep("random", eQTL$gene_CHROM, value = FALSE, invert = T),] # random scaffolds are exluced 
eQTL$class <- factor(eQTL$class, levels = c("cis", "trans"))
eQTL$gene_CHROM <- gsub("(chr)([[:print:]]+)", "\\2", eQTL$gene_CHROM)
eQTL$eQTL_chr <- gsub("(chr)([[:print:]]+)", "\\2", eQTL$eQTL_chr)
eQTL$gene_CHROM <- factor(eQTL$gene_CHROM, levels = c("C09", "C08", "C07", "C06", "C05", "C04", "C03", "C02", "C01", "A10", "A09", "A08", "A07", "A06", "A05", "A04", "A03", "A02", "A01"))
eQTL$eQTL_chr <- factor(eQTL$eQTL_chr, levels = c("A01", "A02", "A03", "A04", "A05", "A06", "A07", "A08", "A09", "A10", "C01", "C02", "C03", "C04", "C05", "C06", "C07", "C08", "C09"))

# plot   
p.eQTL <- 
eQTL %>% 
  ggplot() + 
  geom_point(aes(x = pos, y = gene_start, color = class), size = 0.5) +
  facet_grid(gene_CHROM ~ eQTL_chr, switch = "both", scales = "free") + 
  theme_classic() + 
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) + 
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "eQTL genetic position", y = "gene start") + 
  theme(text = element_text(size=8)) + 
  scale_color_manual(values=c("red", "royalblue"))  

p.eQTL 

ggsave(p.eQTL, filename = "../output/p_eQTL.png", height = 5, width = 7.4) 
```

### trans-eQTL hotspot 
```{r, eval=FALSE}
library(qtlhot)
load("../input/cis_trans_result_new_flipped.Rdata") 

cis_eQTL %>% dim() # 11384    12
trans_eQTL %>% dim() # 15197    12  

length(cis_eQTL$gene_ID[cis_eQTL$gene_ID %in% trans_eQTL$gene_ID]) # 1814 genes not only have cis-eQTL, but also have trans-eQTL 

scanone_eQTL.F2.trans <- scanone_eQTL.F2[,(colnames(scanone_eQTL.F2) %in% c("chr", "pos", trans_eQTL$gene_ID))]

alphas <- seq(0.01, 0.10, by=0.01)  
lod.thr <- lod.thrs[5]
lod.thr 

cross.F2$pheno <- cross.F2$pheno[,(colnames(cross.F2$pheno) %in% c(trans_eQTL$gene_ID))]

### get only significant intervals for each e-trait, using LOD drop method 
high1 <- highlod(scanone_eQTL.F2.trans, lod.thr = min(lod.thrs), drop.lod = 1.5) 
max(high1, lod.thr = lod.thrs) # max number of e-trait fall into loci with different lod threshold   

hots1 <- hotsize(high1, lod.thr = lod.thr)  
summary(hots1) # for each genomic position 

# permutation to get the statistical significance 
# https://github.com/MaloofLab/Li-eQTL-2018/blob/master/scripts/trans_eQTL_hotspot_perm_transonly.R
```

### plot out trans-eQTL result 
```{r}
load("../output/hotperm1.trans.Rdata")
load("../output/hots1.Rdata") 

par(ps = 8, cex = 1, cex.main = 1, cex.axis = 0.8, tcl = -0.3) 
plot(hots1, ylab="trans-eQTL counts") 
abline(h = 129)  
```

### analysis of the trans-eQTL loci for FA composition 

while other poeple all find FAE to be the gene responsible for erucic acid content, I found them to be controlled by a trans-eQTL hotspot on A03. Now I want to see

1. whether adding genotype info of A03 loci info can increase the R2 of erucic acid

2. whether adding expression of the two FAE1 genes can increase the R2 of erucic acid

using brms() package for this purpose

https://github.com/leejimmy93/KIAT/blob/master/F2/scantwo_flipped.Rmd
```{r, eval=FALSE}
library(brms)
load("../input/F2_geno_data_2_Ae_Ol_new.Rdata")

ID <- c("chrC03_53442609", "chrA08_10120555", "chrA03_12781099")
geno <- as.data.frame(t(F2_geno_data_2_Ae_Ol_new[rownames(F2_geno_data_2_Ae_Ol_new) %in% ID,])) 
geno$id <- paste("Sample_F2_", rownames(geno), sep = "")

# expression 
cross.F2 <- read.cross("csvsr", genfile ="../input/LG.f2.madmapper.final_gen_revised_flipped_C05C08.csv", 
                         phefile = "../input/vstMat.f2.batch.corrected_revised.csv",
                         genotypes = c("AA", "AB", "BB"))   

ID_gene <- c("BnaA08g11130D", "BnaC03g65980D")

expression <- cross.F2$pheno[,colnames(cross.F2$pheno) %in% c(ID_gene, "id")] 

# pheno 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv", 
                     genotypes = c("AA", "AB", "BB")) 

pheno <- 
LG.f2.after.crossover$pheno %>% 
  dplyr::select(ends_with("acid"), id)

# combine 
FAE1_data <- 
geno %>% 
  left_join(expression) %>%
  left_join(pheno) 

FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("-",NA,x)))
FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("A",-1,x)))
FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("B",1,x)))
FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("H",0,x)))
FAE1_data$BnaC03g65980D <- as.numeric(FAE1_data$BnaC03g65980D)
FAE1_data$BnaA08g11130D <- as.numeric(FAE1_data$BnaA08g11130D)
FAE1_data$chrA03_12781099 <- as.numeric(as.character(FAE1_data$chrA03_12781099))
FAE1_data$chrA08_10120555 <- as.numeric(as.character(FAE1_data$chrA08_10120555))
FAE1_data$chrC03_53442609 <- as.numeric(as.character(FAE1_data$chrC03_53442609))

FAE1_data$Erucic_acid <- as.numeric(FAE1_data$Erucic_acid)

# stats analysis 
  # just the two loci, with interaction 
  FAE1 <- brm(Erucic_acid ~ chrC03_53442609 * chrA08_10120555,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))  # mean of 0 and sd of 50

  # add genotype from A03, pairwise interaction 
  FAE1.2 <- brm(Erucic_acid ~ chrA03_12781099 + chrA08_10120555 + chrC03_53442609 + chrA03_12781099:chrA08_10120555 + chrA03_12781099:chrC03_53442609 + chrC03_53442609:chrA08_10120555,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))

  # no interaction at all, add genotype from A03 
  FAE1.2.2 <- brm(Erucic_acid ~ chrA03_12781099 + chrA08_10120555 + chrC03_53442609,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))

  # just add interaction between A08 and C03
  FAE1.2.3 <- brm(Erucic_acid ~ chrA03_12781099 + chrA08_10120555 + chrC03_53442609 + chrA08_10120555:chrC03_53442609,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))
  
  ### no interaction, just the two major QTL 
  FAE0 <- brm(Erucic_acid ~ chrC03_53442609 + chrA08_10120555,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))  # mean of 0 and sd of 10 

# Erucic acid A03 is signficant with positive effect from Da-Ol-1 

loo(FAE0, FAE1) 
loo(FAE1.2, FAE1.2.2, FAE1.2.3) 

### hypothesis testing 
# A08 & C03, no interaction 
summary(FAE0) 
hypothesis(FAE0, "chrA08_10120555 < 0")
hypothesis(FAE0, "chrC03_53442609 < 0")  

# only A08 & C03
summary(FAE1)
hypothesis(FAE1, "chrA08_10120555 < 0")
hypothesis(FAE1, "chrC03_53442609 < 0")  

# add A03, pairwise interaction 
summary(FAE1.2)
hypothesis(FAE1.2, "chrA08_10120555 < 0") 
hypothesis(FAE1.2, "chrC03_53442609 < 0") 
hypothesis(FAE1.2, "chrA03_12781099 > 0")

# add A03, no interaction  
summary(FAE1.2.2) 
hypothesis(FAE1.2.2, "chrA08_10120555 < 0")  
hypothesis(FAE1.2.2, "chrC03_53442609 < 0") 
hypothesis(FAE1.2.2, "chrA03_12781099 > 0")  ### allele from Da-Ol-1 increase Erucic acid level. 
plot(hypothesis(FAE1.2.2, "chrA03_12781099 > 0"))   

# add A03, and include interaction between A08 and C03
summary(FAE1.2.3)
hypothesis(FAE1.2.3, "chrA08_10120555 < 0") 
hypothesis(FAE1.2.3, "chrC03_53442609 < 0") 
hypothesis(FAE1.2.3, "chrA03_12781099 > 0")  ### allele from Da-Ol-1 increase Erucic acid level. 
plot(hypothesis(FAE1.2.2, "chrA03_12781099 > 0")) 

# A03 genotype is significant for Erucic acid, how about other FA? test oleic acid, did the same thing, and found the opposite effect for Oleic acid.   
```

### integrate QTL and eQTL mapping result 
```{r}
load("../output/traitQTL.annotated.flipped.Rdata")
load("../output/cis_trans_result_new_flipped_C05C08.Rdata")

cis_eQTL %>% dim() #  11587    15
trans_eQTL %>% dim() # 14657    15  
traitQTL.annotated %>% dim() # 18647    13 

# cis-regulator candidates 
cis_eQTL.qtl.combined <- inner_join(cis_eQTL,traitQTL.annotated,by="gene_ID") 

cis_eQTL.qtl.combined.final <- 
cis_eQTL.qtl.combined %>% 
  mutate(eQTL_start = start.x, eQTL_end = end.x, QTL_start = start.y, QTL_end = end.y, eQTL_genetic_start = genetic_start.x, eQTL_genetic_end = genetic_end.x, QTL_genetic_start = genetic_start.y, QTL_genetic_end = genetic_end.y) %>% 
  dplyr::select(-c(start.x, end.x, start.y, end.y, genetic_start.x, genetic_end.x, genetic_start.y, genetic_end.y)) 

write.csv(cis_eQTL.qtl.combined.final, file = "../output/cis_eQTL.qtl.combined.final.csv")    
  
# trans-eQTL overlap with trait QTL, use genetic interval as the interval for both trait QTL and eQTL  
traitQTL <-  
traitQTL.annotated %>%  
  group_by(trait, chrom, genetic_start, genetic_end, model, start, end) %>% 
  dplyr::summarise(count = n())  

gff.trans_eQTL <- GRanges(seqnames = Rle(trans_eQTL$eQTL_chr),ranges = IRanges(start = trans_eQTL$genetic_start, end = trans_eQTL$genetic_end), names = trans_eQTL$gene_ID) 

gff.trait_qtl<- GRanges(seqnames = Rle(paste("chr", traitQTL$chrom, sep = "")),ranges = IRanges(start = traitQTL$genetic_start, end = traitQTL$genetic_end), trait = traitQTL$trait)   

# trans-regulator target candidates
trans_candidate <- mergeByOverlaps(gff.trait_qtl, gff.trans_eQTL) 
trans_candidate %>% dim()  # 4022    4  

### add the rest of information from trans_eQTL & trait_QTL 
trans_candidate <- 
trans_candidate %>% 
  as.data.frame() %>% 
  mutate(chrom = gff.trait_qtl.seqnames, trait_genetic_start = gff.trait_qtl.start, trait_genetic_end = gff.trait_qtl.end, gene_ID = names, eQTL_chr = gff.trans_eQTL.seqnames, eQTL_genetic_start = gff.trans_eQTL.start, eQTL_genetic_end = gff.trans_eQTL.end) %>% 
  dplyr::select(chrom, trait_genetic_start, trait_genetic_end, trait, gene_ID, eQTL_chr, eQTL_genetic_start, eQTL_genetic_end) 

### annotate trans_eQTL 
load("../input/BnapusAnnotation.Rdata") 

trans_candidate.final <- 
trans_candidate %>%  
  left_join(BnapusAnnotation, c("gene_ID" = "name")) %>% 
  mutate(trait_chrom = chrom.x, gene.chrom = chrom.y, gene.start = start, gene.end = end) %>% 
  dplyr::select(-c(chrom.x, chrom.y, start, end))   

## get GO term for each gene
load("../input/napus_GO_combined.Rdata") 

trans_candidate.final <- 
trans_candidate.final %>% 
  left_join(napus_GO_combined) 

trans_candidate.final %>% dim() # 4022 16 
trans_candidate.final$gene_ID %>% unique() %>% length() # 1972 
write.csv(trans_candidate.final, file = "../output/trans_candidate.final.csv")  
```

### cis-coding candidates  
```{r, eval=FALSE}
SNP_all <-  read.csv("../input/vcf.Ae.Ol.intersect.df.2.csv", row.names = 1, as.is = T)

SNP_all <- data.frame(sapply(SNP_all, function(i) sub("1-Jan", "1/1", i))) %>% 
  dplyr::select(-subgenome, -Chr_ID)

write.csv(SNP_all, file = "../output/SNP_all.csv")

# sh ~/KIAT/F2/SNP_info_extract_vcf_ann.sh ~/F2/for_paper/candidate_gene_SNP_mutation/SNP_all.csv /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/SNP_parent/result/freebayes/Ae_Ol_modified_SNPonly_biallelic.recode.ann.modified2.vcf 

sig_SNP_ann <- read_table2("../input/significant_SNP_annotation", col_names=F)
dim(sig_SNP_ann) # 62875     6 
colnames(sig_SNP_ann) <- c("CHROM", "POS", "REF", "ALT", "mutation_type", "impact")

SNP_all$POS <- SNP_all$POS %>% as.character() %>% as.numeric()

SNP_all <- 
SNP_all %>% 
  left_join(sig_SNP_ann)

gff.mRNA
colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 
genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), names = gff.mRNA$name)

SNP <- GRanges(seqnames = Rle(sig_SNP_ann$CHROM), ranges = IRanges(start = sig_SNP_ann$POS, end = sig_SNP_ann$POS), ID = paste(sig_SNP_ann$CHROM, sig_SNP_ann$POS, sep = "_"))

SNP_gene <- mergeByOverlaps(SNP, genes) # warning message here. look out!!!!!! 

SNP_gene_df <- as.data.frame(SNP_gene)
SNP_gene_df %>% dim() # 58969    14

SNP_gene_final <- SNP_gene_df[,c("SNP.ID", "genes.seqnames", "SNP.start", "names")]
SNP_all$SNP.ID <- paste(SNP_all$CHROM, SNP_all$POS, sep = "_") 

SNP_all <- 
SNP_all %>% 
  left_join(SNP_gene_final)

dim(SNP_all) #  62958    12

unique(SNP_all$mutation_type) 

# the kind of mutations that could affect protein function here: "missense_variant&splice_region_variant", "missense_variant", "stop_lost", "splice_region_variant&synonymous_variant", "splice_donor_variant&intron_variant", "splice_region_variant&intron_variant", "stop_gained&splice_region_variant", "splice_acceptor_variant&intron_variant", "stop_lost&splice_region_variant", "splice_region_variant&stop_retained_variant", "initiator_codon_variant", "start_lost"  

# combine mutation type with candidate genes 
load("../output/traitQTL.annotated.flipped.Rdata")

traitQTL.annotated <- 
traitQTL.annotated %>% 
  semi_join(SNP_all, c("gene_ID" = "names")) %>% # keep all genes which have SNP annotations 
  left_join(SNP_all, c("gene_ID" = "names"))   # add SNP info/annotation to the QTL gene

# just keep mutations which would affect protein function 
traitQTL.annotated.SNP <- 
traitQTL.annotated %>% 
  filter(mutation_type != "synonymous_variant" & 
           mutation_type != "downstream_gene_variant" & 
           mutation_type != "upstream_gene_variant" & 
           mutation_type != "intron_variant" & 
           mutation_type != "stop_retained_variant") %>% 
  dplyr::select(trait, SNP.ID, mutation_type, impact, gene_ID, AGI, At_symbol, B.napus_GO_term, GO_des, chrom, Ae.gt, Ol.gt, REF, ALT) 
 
write.csv(traitQTL.annotated.SNP, file = "../output/traitQTL.annotated.SNP.csv")
```

### plot for three genes 
```{r}
load("../output/cim.qtl_two_genes.Rdata")
load("../output/cim.qtl_FLC.Rdata")
source("../scripts/helper.R") 

oil_eQTL <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["Erucic_acid"]], method = "Erucic_acid", group = "QTL"), 
                       data.frame(cim.qtl[["Oleic_acid"]], method = "Oleic_acid", group = "QTL"), 
                       data.frame(cim.qtl[["Stearic_acid"]], method = "Stearic_acid", group = "QTL"),
                       data.frame(cim.qtl[["Palmitic_acid"]], method = "Palmitic_acid", group = "QTL"),
                       data.frame(cim.qtl[["Linoleic_acid"]], method = "Linoleic_acid", group = "QTL"),
                       data.frame(cim.qtl[["Linolenic_acid"]], method = "Linolenic_acid", group = "QTL"),
                       data.frame(cim.qtl[["vaccenic_acid"]], method = "vaccenic_acid", group = "QTL"),
                       data.frame(cim.qtl[["BnaA08g11140D"]], method = "BnaA08g11140D", group = "eQTL"),
                       data.frame(cim.qtl[["BnaA08g11060D"]], method = "BnaA08g11060D", group = "eQTL")
                       ),
         chrs = c("A08"), 
         # lod = cim.perm[[1]], 
         title = "", 
         rug = TRUE)     

growth_eQTL <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["days_to_bolt"]], method = "days_to_bolt", group = "QTL"), 
                       data.frame(cim.qtl[["days_to_flower"]], method = "days_to_flower", group = "QTL"), 
                       data.frame(cim_growth_model_trait.F2[["height_Hmax"]], method = "height_Hmax", group = "QTL"), 
                       data.frame(cim_growth_model_trait.F2[["leaf_number_I"]], method = "leaf_number_I", group = "QTL"), 
                       data.frame(cim_growth_model_trait.F2[["BnaA10g22080D"]], method = "BnaA10g22080D", group = "eQTL"),     
                       data.frame(cim.qtl[["root_weight_2016_05_13"]], method = "root_weight_2016_05_13", group = "QTL")), 
                       
         chrs = c("A10"), 
         # lod = cim.perm[[1]], 
         title = "", 
         rug = TRUE)       
```

